ezmlm-split.1
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
.TH ezmlm-split 1
.SH NAME
ezmlm-split \- distribute (un)subscribe requests to sublists
.SH SYNOPSIS
.B ezmlm-split
.I dir
[
.B \-dD
][
.I splitfile
]
.SH DESCRIPTION
If the action is
.I \-subscribe
or
.IR \-unsubscribe ,
.B ezmlm-split
analyzes the target of the request,
 computing a hash in the range 0-52 from
the address, and determines the ``domain key'', i.e. 
the two top levels of the host address in reverse order.
Thus, the domain entry for ``d@a.b.c'' becomes ``c.b''
and the one for ``d@a'' becomes ``a''.

The hash and domain parts are then tested against successive lines of
.I splitfile
(default is
.IR dir\fB/split ).
If a match is found, the request is forwarded to the corresponding
sublist, and
.B ezmlm-split
exits 99. If a match is not found or action is not
.I \-subscribe
or
.IR \-unsubscribe ,
.B ezmlm-split
exits 0.

In
.IR splitfile ,
blank lines and lines with ``#'' in position 1 are ignored. Other lines are
expected to be of the format ``dom:low:hi:list@host'', where ``dom'' is
the top level domain, ``low''-``hi'' the range of the hash (defaults 0 and 52),
and ``list@host'' the name of the corresponding list (default is the
local list). A line is considered to match if the address hash is
between ``low'' and ``hi'' inclusive and ``dom'' is empty,
or if the ``domain key'' matches ``dom'' for the full length of ``dom''. Thus,
the address ``user@aol.com'' would match ``com'' and ``aol.com'',
but not ``host.com''.

If the domain
specified is the top level domain up to 3 characters, the split is identical
as that used by the SQL subscriber interface. This is recommended.
There can be several entries for a given sublist.

.B ezmlm-split
can be used also for list with SQL-based sublisting. In this case,
addresses matching the
.I splitfile
are forwarded to the respective non-SQL sublist, and non-matching addresses
are handed by the SQL sublist.


If
.I dir\fB/qmqpservers
exists,
.B ezmlm-split will use
.B qmail-qmqp(1)
to send messages.
.SH OPTIONS
.TP
.B \-d
(Default.)
Do. Forward requests to the appropriate list.
.TP
.B \-D
Do not process messages. Rather, read addresses, one per line from stdin, and
print ``sublist@host: address'' where ``sublist@host'' is the address to which
the request would have been forwarded in normal operation. This is used to
process a set of existing addresses into a set of address collections, one
per sublist. The output can be sorted and easily processed into a set of files,
one per sublist containing the addresses that sublist handles.
.SH "SPLIT LIST SETUP"
To use a hierarchy of sublists, set up the master list and add a
.B ezmlm-split
line before the
.B ezmlm-manage(1)
line in
.IR dir\fB/manager .
Create any number of sublists of the main list on other local or
distant hosts. Ideally, these should be non-archived, to point to the correct
message numbers of the main list archive (see
.BR ezmlm-send(1) .
You can use
.B ezmlm-make -C\fIezmlmsubrc
for this. If you don't, use the text files from the main list, except
.IR bounce-bottom .
Next, create
.I split
in the main list directory to achieve an appropriate split. Load splitting
between several local hosts is best achieved by hash, with a blank domain.
Geographical splitting with hosts in different countries is best done
via ``domain'' and naturally, a large domain can be subdivided by hash.

Subscribers will receive all messages 'From:' the main list. When they
subscribe or unsubscribe, the request will be forwarded to the appropriate
sublist, which will handle the confirmation. All information, except
.I bounce-bottom
refers the user to the main list. Thus, to the user the list appears as
a single list with the address of the main list, and distribution among
sublists is at the discretion of the administrator of the main list.

.SH "ADDING/REMOVING SUBLISTS"
In general, the main list should be disabled, while changing the sublist
split. This can be done by changing the mode of
.I dir\fB/lock
to 0 or by setting the sticky bit for the home directory of the user
controlling the list.

To remove a sublist, edit the lines for that sublist in the splitfile to
point to another list, and add the existing subscribers of the removed
sublist to the sublist taking the load.
When splitting a sublist into several sublists, create the new sublists,
and edit the split file to distribute the load
between them (usually by hash). Process the subscribers of the old list
with:

.Vb 1
 | ezmlm-split -D dir | sort | program
.Ve
where to get one file of addresses per new sublist, ``program'' could be:

.Vb 12
\&#!/usr/bin/perl
\&while (<>) {
\&  ($f,$t) = split (':');
\&  $t =~ s/^\ //;
\&  if ($f ne $of) {
\&    $of = $f;
\&    close(OUT) if ($of);
\&    open(OUT,">$f") or die("Unable to open $f");
\&  }
\&  print OUT $t;
\&}
\&close(OUT) if ($of);
.Ve

Remove all subscribers from the old list,
and add the respective subscribers to each list.

For any more drastic reorganizations, collect all the subscribes of the
affected sublists, create the new subscriber lists, and update the
subscribers of the affected lists.

Reorganizations are easier done when lists use SQL support. By
temporarily using SQL support, reorganizations can be done on running
lists even when normally using
.B ezmlm-split
and local subscriber databases.
.SH "SEE ALSO"
ezmlm-list(1),
ezmlm-make(1),
ezmlm-manage(1),
ezmlm-sub(1)
ezmlm-unsub(1),
ezmlm(5),
ezmlmrc(5),
ezmlmsubrc(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-split.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "error.h"
#include "stralloc.h"
#include "str.h"
#include "env.h"
#include "sig.h"
#include "open.h"
#include "scan.h"
#include "strerr.h"
#include "byte.h"
#include "getln.h"
#include "case.h"
#include "qmail.h"
#include "sgetopt.h"
#include "subfd.h"
#include "substdio.h"
#include "readwrite.h"
#include "quote.h"
#include "now.h"
#include "uint32.h"
#include "subhash.h"
#include "fmt.h"
#include "messages.h"
#include "die.h"
#include "config.h"
#include "idx.h"
#include "auto_version.h"

const char FATAL[] = "ezmlm-split: fatal: ";
const char INFO[] = "ezmlm-split: info: ";
const char USAGE[] =
"ezmlm-split: usage: ezmlm-split [-dD] dir [splitfile]";

int flagdo = 1;		/* default is manager function */

const char *sender;
const char *split;
stralloc target = {0};
stralloc lctarget = {0};
stralloc line = {0};
stralloc domain = {0};
stralloc name = {0};
stralloc from = {0};
stralloc to = {0};
char strnum[FMT_ULONG];
unsigned long lineno;
int flagfound;

void die_syntax(void)
{
  strnum[fmt_ulong(strnum,lineno)] = '\0';
  strerr_die6x(111,FATAL,split," syntax error line ",strnum,": ",line.s);
}

char spbuf[1024];	/* should normally hold entire file */
substdio sssp;

struct qmail qq;

int findname(void)
/* returns 1 if a matching line was found, 0 otherwise. name will contain */
/* the correct list address in either case */
{
  char *cpat,*cp1,*cp2,*cplast;
  const char *cpname;
  unsigned long u;
  unsigned char hash,hash_hi,hash_lo;
  unsigned int pos,pos_name,pos_hi;
  int fd,match;

  /* make case insensitive hash */
  flagfound = 0;			/* default */
  cpname = "";				/* default */
  if (!stralloc_copy(&lctarget,&target)) die_nomem();
  case_lowerb(lctarget.s,lctarget.len -1);
  hash = subhashs(lctarget.s);

  /* make domain pointer */
  cpat = lctarget.s + str_chr(lctarget.s,'@');
  if (!*cpat)
    strerr_die4x(100,FATAL,MSG(ERR_ADDR_AT),": ",target.s);
  cplast = cpat + str_len(cpat) - 1;
  if (*cplast == '.') --cplast;		/* annonying special case */
  cp1 = cpat + byte_rchr(cpat,cplast - cpat, '.');
  if (cp1 != cplast) {			/* got one '.' */
    if (!stralloc_copyb(&domain,cp1 + 1, cplast - cp1)) die_nomem();
    cp2 = cpat + byte_rchr(cpat, cp1 - cpat,'.');
    if (cp2 == cp1) cp2 = cpat;
    ++cp2;
    if (!stralloc_append(&domain,".")) die_nomem();
    if (!stralloc_catb(&domain,cp2, cp1 - cp2)) die_nomem();
  } else				/* no '.' */
    if (!stralloc_copyb(&domain,cpat + 1,cplast - cpat)) die_nomem();
  if (!stralloc_0(&domain)) die_nomem();

  if ((fd = open_read(split)) == -1)
    strerr_die2sys(111,FATAL,MSG1(ERR_OPEN,split));
  substdio_fdbuf(&sssp,read,fd,spbuf,(int) sizeof(spbuf));
  lineno = 0;
  for (;;) {	/* dom:hash_lo:hash_hi:listaddress */
    if (getln(&sssp,&line,&match,'\n') == -1)
      strerr_die2sys(111,FATAL,MSG1(ERR_READ,split));
     lineno++;
    if (!match)
      break;
    if (line.s[0] == '#') continue;	/* comment */
    line.s[line.len - 1] = '\0';	/* no need to allow \0 in lines */
    if (!line.s[pos = str_chr(line.s,':')])
      continue;				/* usually blank line */
    line.s[pos] = '\0';
    if (pos == 0 ||			/* no domain */
	  (case_starts(domain.s,line.s))) {	/* or matching domain */
	if (!line.s[++pos]) die_syntax();
	pos_hi = pos + str_chr(line.s + pos,':');
	if (!line.s[pos_hi]) die_syntax();
	pos_hi++;
	(void) scan_ulong(line.s + pos, &u);	/* scan_uint() not in ezmlm */
	hash_lo = (unsigned char) u;
	(void) scan_ulong(line.s + pos_hi, &u);
	hash_hi = (unsigned char) u;
	pos_name = pos_hi + str_chr(line.s + pos_hi,':');
	if (pos_hi == pos_name) hash_hi = 52L;	/* default hi = 52 */
	if (line.s[pos_name]) pos_name++;
	if (hash > hash_hi || hash < hash_lo) continue;	/* not us */
	cpname = line.s + pos_name;
	while (*cpname &&				/* isolate name */
	    (*cpname == ' ' || *cpname == '\t')) cpname++;
	pos = line.len - 2;
	while (pos && (line.s[pos] == '\n' || line.s[pos] == ' ' ||
		line.s[pos] == '\t')) line.s[pos--] = '\0';
	break;
    }
  }
  close(fd);

  if (*cpname) {
    if (!stralloc_copys(&name,cpname)) die_nomem();
    if (byte_chr(name.s,name.len,'@') == name.len) {	/* local sublist */
      if (!stralloc_append(&name,"@")) die_nomem();
      if (!stralloc_cat(&name,&outhost)) die_nomem();
    }
    if (!stralloc_0(&name)) die_nomem();
    return 1;
  } else {			/* match without name or no match =>this list */
    if (!stralloc_copy(&name,&outlocal)) die_nomem();
    if (!stralloc_append(&name,"@")) die_nomem();
    if (!stralloc_cat(&name,&outhost)) die_nomem();
    if (!stralloc_0(&name)) die_nomem();
    return 0;
  }
}

void main(int argc,char **argv)
{
  char *dir;
  char *action;
  char *dtline;
  char *nhost;
  const char *err;
  unsigned int i;
  int match;
  int opt;

  sig_pipeignore();

  while ((opt = getopt(argc,argv,"dDvV")) != opteof) {
    switch (opt) {
    case 'd': flagdo = 1; break;
    case 'D': flagdo = 0; break;
    case 'v':
    case 'V':
      strerr_die2x(0, "ezmlm-split version: ",auto_version);
    default:
      die_usage();
    }
  }

  startup(dir = argv[optind++]);
  if (!(split = argv[optind]))
    split = "split";

  if (flagdo) {
    sender = env_get("SENDER");
    if (!sender) strerr_die2x(100,FATAL,MSG(ERR_NOSENDER));
    if (!*sender)
      strerr_die2x(100,FATAL,MSG(ERR_BOUNCE));
    if (!sender[str_chr(sender,'@')])
      strerr_die2x(100,FATAL,MSG(ERR_ANONYMOUS));
    if (str_equal(sender,"#@[]"))
      strerr_die2x(100,FATAL,MSG(ERR_BOUNCE));

    action = env_get("DEFAULT");
    if (!action) strerr_die2x(100,FATAL,MSG(ERR_NODEFAULT));
    if (!stralloc_copys(&target,sender)) die_nomem();
    if (action[0]) {
      i = str_chr(action,'-');
      if (action[i]) {
        action[i] = '\0';
        if (!stralloc_copys(&target,action + i + 1)) die_nomem();
        i = byte_rchr(target.s,target.len,'=');
        if (i < target.len)
	  target.s[i] = '@';
      }
    }
    if (!stralloc_0(&target)) die_nomem();

    if (case_diffs(action,ACTION_SUBSCRIBE) &&
      case_diffs(action,ALT_SUBSCRIBE) &&
      case_diffs(action,ACTION_UNSUBSCRIBE) &&
      case_diffs(action,ALT_UNSUBSCRIBE))
    _exit(0);			/* not for us */

    if (findname()) {
				/* new sender */
      if (!stralloc_copy(&from,&outlocal)) die_nomem();
      if (!stralloc_cats(&from,"-return-@")) die_nomem();
      if (!stralloc_cat(&from,&outhost)) die_nomem();
      if (!stralloc_0(&from)) die_nomem();
      nhost = name.s + str_rchr(name.s,'@');		/* name must have '@'*/
      *(nhost++) = '\0';
      if (!stralloc_copys(&to,name.s)) die_nomem();	/* local */
      if (!stralloc_append(&to,"-")) die_nomem();	/* - */
      if (!stralloc_cats(&to,action)) die_nomem();	/* subscribe */
      if (!stralloc_append(&to,"-")) die_nomem();	/* - */
      if (target.s[i = str_rchr(target.s,'@')])
	target.s[i] = '=';
      if (!stralloc_cats(&to,target.s)) die_nomem();	/* target */
      if (!stralloc_append(&to,"@")) die_nomem();	/* - */
      if (!stralloc_cats(&to,nhost)) die_nomem();	/* host */
      if (!stralloc_0(&to)) die_nomem();
      dtline = env_get("DTLINE");
      if (!dtline) strerr_die2x(100,FATAL,MSG(ERR_NODTLINE));

      if (qmail_open(&qq,(stralloc *) 0) == -1)
        strerr_die2sys(111,FATAL,MSG(ERR_QMAIL_QUEUE));
      qmail_puts(&qq,dtline);				/* delivered-to */
      if (qmail_copy(&qq,subfdin,0) != 0)
        strerr_die2sys(111,FATAL,MSG(ERR_READ_INPUT));
      qmail_from(&qq,from.s);
      qmail_to(&qq,to.s);

      if (*(err = qmail_close(&qq)) != '\0')
        strerr_die4x(111,FATAL,MSG(ERR_TMP_QMAIL_QUEUE),": ",err + 1);

      strnum[fmt_ulong(strnum,qmail_qp(&qq))] = 0;
      strerr_die3x(99,INFO,"qp ",strnum);
    }
    _exit(0);
  } else {

    for (;;) {
      if (getln(subfdin,&line,&match,'\n') == -1)
	  strerr_die2sys(111,FATAL,MSG(ERR_READ_INPUT));
      if (!match) break;
      if (line.len == 1) continue;	/* ignore blank lines */
      if (line.s[0] == '#') continue;	/* ignore comments */
      if (!stralloc_copy(&target,&line)) die_nomem();
      target.s[target.len - 1] = '\0';
      (void) findname();
      if (!stralloc_cats(&name,": ")) die_nomem();
      if (!stralloc_cats(&name,target.s)) die_nomem();
      if (!stralloc_append(&name,"\n")) die_nomem();
      if (substdio_put(subfdout,name.s,name.len) == -1)
	strerr_die2sys(111,FATAL,MSG(ERR_WRITE_STDOUT));
    }
    if (substdio_flush(subfdout) == -1)
      strerr_die2sys(111,FATAL,MSG(ERR_FLUSH_STDOUT));
    _exit(0);
  }
  (void)argc;
}
